### ISSUE TITLE
`std_instead_of_core` does not detect uses of non-fully qualified items

### ISSUE THREAD

**Issue by japaric** [rust-lang/rust-clippy#11159]
### Summary

`std_instead_of_core` does not detect the use of, for example, the `std::fmt::Display` trait when it appears as `fmt::Display` accompanied by `use std::fmt`

### Lint Name

clippy::std_instead_of_core

### Reproducer

I tried this code:

```rust
#![deny(clippy::std_instead_of_core)]
use std::fmt;
struct S;
impl fmt::Display for S {
    fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result {
        todo!()
    }
}
```

*I expected to see this happen:* 3 warnings on `fmt::Display`, `fmt::Formatter` and `fmt::Result`

*Instead, this happened:* no warnings were produced

---

The lint is able to detect imported items and items qualified by absolute path:
``` rust
#![deny(clippy::std_instead_of_core)]
use std::fmt::{self, Display};
struct S;
impl Display for S {
    fn fmt(&self, _: &mut std::fmt::Formatter<'_>) -> fmt::Result {
        todo!()
    }
}
```
in this second version, both `Display` and `Formatter` are detected but `fmt::Result` is not.

### Version

```text
rustc 1.73.0-nightly (7bd81ee19 2023-07-13)
binary: rustc
commit-hash: 7bd81ee1902c049691d0a1f03be5558bee51d100
commit-date: 2023-07-13
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
This would be tricky to fix, as reexported items' `res` essentially 'skip' the reexport, giving you `core::fmt::Formatter` here. I'm not sure what the best way of going about this would be.

**Comment by Jarcho** [comment]
It should be noted that `std::fmt` is not the same module as `core::fmt`, but an alias of `alloc::fmt`. Linting this would require verifying that none of the additional items provided are referenced through the import.

There is still a bug here as `alloc_instead_of_core` should trigger, but doesn't on this case.

**Comment by bushrat011899** [comment]
An added complication is re-exports. `pub use std::fmt;` can and should suggest replacement with `alloc::fmt`, but it could never suggest replacing it with `core::fmt`, since that would be a semver breaking change.
